client/v3/snapshot/v3_snapshot.go

```diff
@@ -36,13 +36,81 @@ func hasChecksum(n int64) bool {
 	return (n % 512) == sha256.Size
 }
 
+// SaveWithVersion fetches snapshot from remote etcd server, saves data
+// to target path and returns server version. If the context "ctx" is canceled or timed out,
+// snapshot save stream will error out (e.g. context.Canceled,
+// context.DeadlineExceeded). Make sure to specify only one endpoint
+// in client configuration. Snapshot API must be requested to a
+// selected node, and saved snapshot is the point-in-time state of
+// the selected node.
+// Supported since etcd 3.6.
+func SaveWithVersion(ctx context.Context, lg *zap.Logger, cfg clientv3.Config, dbPath string) (version string, err error) {
+	if lg == nil {
+		lg = zap.NewExample()
+	}
+	cfg.Logger = lg.Named("client")
+	if len(cfg.Endpoints) != 1 {
+		return "", fmt.Errorf("snapshot must be requested to one selected node, not multiple %v", cfg.Endpoints)
+	}
+	cli, err := clientv3.New(cfg)
+	if err != nil {
+		return "", err
+	}
+	defer cli.Close()
+
+	partpath := dbPath + ".part"
+	defer os.RemoveAll(partpath)
+
+	var f *os.File
+	f, err = os.OpenFile(partpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fileutil.PrivateFileMode)
+	if err != nil {
+		return "", fmt.Errorf("could not open %s (%v)", partpath, err)
+	}
+	lg.Info("created temporary db file", zap.String("path", partpath))
+
+	now := time.Now()
+	resp, err := cli.SnapshotWithVersion(ctx)
+	if err != nil {
+		return resp.Version, err
+	}
+	defer resp.Snapshot.Close()
+	lg.Info("fetching snapshot", zap.String("endpoint", cfg.Endpoints[0]))
+	var size int64
+	size, err = io.Copy(f, resp.Snapshot)
+	if err != nil {
+		return resp.Version, err
+	}
+	if !hasChecksum(size) {
+		return resp.Version, fmt.Errorf("sha256 checksum not found [bytes: %d]", size)
+	}
+	if err = fileutil.Fsync(f); err != nil {
+		return resp.Version, err
+	}
+	if err = f.Close(); err != nil {
+		return resp.Version, err
+	}
+	lg.Info("fetched snapshot",
+		zap.String("endpoint", cfg.Endpoints[0]),
+		zap.String("size", humanize.Bytes(uint64(size))),
+		zap.String("took", humanize.Time(now)),
+		zap.String("etcd-version", version),
+	)
+
+	if err = os.Rename(partpath, dbPath); err != nil {
+		return resp.Version, fmt.Errorf("could not rename %s to %s (%v)", partpath, dbPath, err)
+	}
+	lg.Info("saved", zap.String("path", dbPath))
+	return resp.Version, nil
+}
+
 // Save fetches snapshot from remote etcd server and saves data
 // to target path. If the context "ctx" is canceled or timed out,
 // snapshot save stream will error out (e.g. context.Canceled,
 // context.DeadlineExceeded). Make sure to specify only one endpoint
 // in client configuration. Snapshot API must be requested to a
 // selected node, and saved snapshot is the point-in-time state of
 // the selected node.
+// Deprecated: Use SaveWithVersion instead.
 func Save(ctx context.Context, lg *zap.Logger, cfg clientv3.Config, dbPath string) error {
 	if lg == nil {
```

Comment:
Can this code call: SaveWithVersion ?

Reply:
Yes, forgot to revert it when debugging SnapshotWithVersion.